src/SMatrix.jl

```diff
@@ -52,6 +52,34 @@ end
     end
 end
 
+@generated function SMatrix{M, N, T}(gen::Base.Generator) where {M, N, T}
```

Comment:
What should we do when it is a `Base.HasShape` iterator and the `size` is different?

Reply:
I would say that the result should be the same as if the generator was first `collect`ed into an array, and this array then passed to the constructor.

Reply:
That sounds reasonable.  (But I didn't know this "works"...)

```julia
julia> SMatrix{2, 3}(ones(3, 2))
2×3 SArray{Tuple{2,3},Float64,2,6} with indices SOneTo(2)×SOneTo(3):
 1.0  1.0  1.0
 1.0  1.0  1.0
```